File: ts/src/graph.ts
import type { Context, RuleFunc, EvalOption, EvalResult, Specs, TraceEntry, NodeEvent, NodeHandler, RunResult } from "./types.js";
import { Strength, EvalMethod, NodeEventType } from "./types.js";
import type { DefeatEdge } from "./defeat-edge.js";
import type { RuleMeta } from "./rule-meta.js";
import { Rule } from "./rule.js";
import { ruleID } from "./rule-id.js";
import { shortName } from "./short-name.js";
import { detectCycle } from "./detect-cycle.js";
import { createRunView, buildAttackerEvents } from "./run-view.js";

export class Graph {
  readonly name: string;
  rules: RuleMeta[] = [];
  roles = new Map<string, string>();
  defeats: DefeatEdge[] = [];

  constructor(name: string) { this.name = name; }

  rule(fn: RuleFunc): Rule { return this._register(fn, Strength.Defeasible, "rule"); }
  counter(fn: RuleFunc): Rule { return this._register(fn, Strength.Defeasible, "counter"); }
  except(fn: RuleFunc): Rule { return this._register(fn, Strength.Defeater, "except"); }

  private _register(fn: RuleFunc, strength: Strength, role: string): Rule {
    const id = ruleID(fn, []);
    if (this.roles.has(id)) throw new Error(`duplicate rule registration: ${id}`);
    const idx = this.rules.length;
    this.rules.push({ name: id, qualifier: 1.0, strength, defeats: [], specs: [], fn });
    this.roles.set(id, role);
    return new Rule(id, this, idx, fn);
  }

  evaluate(ctx: Context, option?: EvalOption): EvalResult[] {
    if (ctx == null) throw new Error("ctx must not be null");
    return this._evaluate(ctx, resolveOption(option), false).results;
  }

  run(ctx: Context, option?: EvalOption): RunResult {
    if (ctx == null) throw new Error("ctx must not be null");
    const opt = resolveOption(option);
    opt.trace = false;      // 강제 off — full pass는 공유·비초기화 상태에서만 정확
    opt.duration = false;
    const st = this._evaluate(ctx, opt, true);   // full pass

    // 디스패치 전 1회 — 불변 스냅샷
    const order: NodeEvent[] = [];
    const byName = new Map<string, NodeEvent>();
    for (const r of this.rules) {
      const type = classifyEvent(st.active.get(r.name) ?? false, st.verdictCache.get(r.name) ?? 0);
      const ne: NodeEvent = {
        name: shortName(r.name),
        role: this.roles.get(r.name) ?? "rule",
        type,
        verdict: st.verdictCache.get(r.name) ?? 0,
        evidence: st.evidence.get(r.name),
      };
      order.push(ne);
      byName.set(ne.name, ne);
    }
    const attackers = buildAttackerEvents(this.defeats, byName);
    const view = createRunView(order, byName, attackers);

    for (let i = 0; i < order.length; i++) {
      const ne = order[i];
      const meta = this.rules[i];          // 인덱스 대응 — shortName 매핑 불필요
      const h = selectHandler(meta, ne.type);
      if (!h) continue;
      try {
        h(ctx, ne, view);
      } catch (e) {
        throw new Error(`handler "${shortName(meta.name)}" (${NodeEventType[ne.type]}): ${e}`);
      }
    }
    return { results: st.results, view };
  }

  private _evaluate(ctx: Context, opt: { method: EvalMethod; trace: boolean; duration: boolean }, full: boolean): {
    results: EvalResult[];
    active: Map<string, boolean>;
    verdictCache: Map<string, number>;
    evidence: Map<string, unknown>;
  } {
    // Build eval context
    const fnMap = new Map<string, RuleFunc>();
    const qualMap = new Map<string, number>();
    const strMap = new Map<string, Strength>();
    const specsMap = new Map<string, Specs>();
    const edges = new Map<string, string[]>();
    const roleMap = new Map(this.roles);

    for (const r of this.rules) {
      fnMap.set(r.name, r.fn);
      qualMap.set(r.name, r.qualifier);
      strMap.set(r.name, r.strength);
      specsMap.set(r.name, r.specs);
    }
    for (const d of this.defeats) {
      const list = edges.get(d.to) ?? [];
      list.push(d.from);
      edges.set(d.to, list);
    }

    // Cycle detection
    const cycleErr = detectCycle(edges);
    if (cycleErr) throw cycleErr;

    // Build attacker set
    const attackerSet = new Set<string>();
    for (const attackers of edges.values()) {
      for (const a of attackers) attackerSet.add(a);
    }

    // Evaluation state
    const ran = new Set<string>();
    const active = new Map<string, boolean>();
    const evidence = new Map<string, unknown>();
    const verdictCache = new Map<string, number>();
    let trace: TraceEntry[] = [];
    let err: Error | null = null;

    function isWarrant(name: string): boolean {
      const str = strMap.get(name) ?? Strength.Defeasible;
      if (str === Strength.Defeater) return false;
      return !attackerSet.has(name);
    }

    function inferRole(id: string): string {
      if (strMap.get(id) === Strength.Defeater) return "except";
      if (attackerSet.has(id)) return "counter";
      return "rule";
    }

    function calc(id: string): number {
      if (err) return -1.0;
      const cached = verdictCache.get(id);
      if (cached !== undefined) return cached;
      const fn = fnMap.get(id);
      if (!fn) return -1.0;
      if (!ran.has(id)) {
        ran.add(id);
        try {
          const [act, ev] = fn(ctx, specsMap.get(id) ?? []);
          active.set(id, act);
          evidence.set(id, ev);
        } catch (e) {
          err = new Error(`rule "${id}": ${e}`);
          return -1.0;
        }
      }
      if (!active.get(id)) return -1.0;
      let sum = 0;
      if (strMap.get(id) !== Strength.Strict) {
        for (const aid of edges.get(id) ?? []) {
          sum += (calc(aid) + 1.0) / 2.0;
        }
      }
      const q = qualMap.get(id) ?? 1.0;
      const v = 2 * (q / (1 + sum)) - 1;
      verdictCache.set(id, v);
      return v;
    }

    function calcTrace(id: string, withDuration: boolean): number {
      if (err) return -1.0;
      const cached = verdictCache.get(id);
      if (cached !== undefined) return cached;
      const fn = fnMap.get(id);
      if (!fn) return -1.0;
      if (!ran.has(id)) {
        ran.add(id);
        const start = withDuration ? performance.now() : 0;
        try {
          const [act, ev] = fn(ctx, specsMap.get(id) ?? []);
          active.set(id, act);
          evidence.set(id, ev);
        } catch (e) {
          err = new Error(`rule "${id}": ${e}`);
          return -1.0;
        }
        const elapsed = withDuration ? performance.now() - start : undefined;
        trace.push({
          name: id,
          role: inferRole(id),
          activated: active.get(id) ?? false,
          qualifier: qualMap.get(id) ?? 1.0,
          evidence: evidence.get(id),
          specs: specsMap.get(id),
          duration: elapsed,
        });
      }
      if (!active.get(id)) return -1.0;
      let sum = 0;
      if (strMap.get(id) !== Strength.Strict) {
        for (const aid of edges.get(id) ?? []) {
          sum += (calcTrace(aid, withDuration) + 1.0) / 2.0;
        }
      }
      const q = qualMap.get(id) ?? 1.0;
      const v = 2 * (q / (1 + sum)) - 1;
      verdictCache.set(id, v);
      return v;
    }

    function reset(): void {
      ran.clear();
      active.clear();
      evidence.clear();
      verdictCache.clear();
      trace = [];
      err = null;
    }

    // Main evaluation loop
    const results: EvalResult[] = [];
    for (const r of this.rules) {
      if (!isWarrant(r.name)) continue;
      if (opt.trace) reset();

      const verdict = opt.trace
        ? calcTrace(r.name, opt.duration)
        : calc(r.name);

      if (err) throw err;
      if (!active.get(r.name)) continue;

      const result: EvalResult = {
        name: shortName(r.name),
        verdict,
        evidence: evidence.get(r.name),
      };
      if (opt.trace) {
        result.trace = trace.map(t => ({ ...t, name: shortName(t.name) }));
      }
      results.push(result);
    }

    if (full) {                                // run은 trace를 강제 off하므로 calc(non-trace)만 탐
      for (const r of this.rules) {
        if (!ran.has(r.name)) calc(r.name);    // 미도달 노드까지 평가
      }
      if (err) throw err;
    }

    return { results, active, verdictCache, evidence };
  }
}

function classifyEvent(active: boolean, verdict: number): NodeEventType {
  if (!active) return NodeEventType.Inactive;
  return verdict > 0 ? NodeEventType.Active : NodeEventType.Defeated;
}

function selectHandler(r: RuleMeta, t: NodeEventType): NodeHandler | undefined {
  switch (t) {
    case NodeEventType.Active: return r.onActive;
    case NodeEventType.Defeated: return r.onDefeated;
    default: return r.onInactive;
  }
}

function resolveOption(opt?: EvalOption): { method: EvalMethod; trace: boolean; duration: boolean } {
  const resolved = {
    method: opt?.method ?? EvalMethod.Matrix,
    trace: opt?.trace ?? false,
    duration: opt?.duration ?? false,
  };
  if (resolved.duration) resolved.trace = true;
  return resolved;
}

File: py/src/rulecat/graph.py
from __future__ import annotations
import time
from typing import Any

from rulecat.types import (
    Context, EvalMethod, EvalOption, EvalResult, Strength, TraceEntry,
    RunResult,
)
from rulecat.trace import Trace
from rulecat.defeat_edge import DefeatEdge
from rulecat.rule_meta import RuleMeta
from rulecat.rule import Rule
from rulecat.rule_id import rule_id
from rulecat.short_name import short_name
from rulecat.detect_cycle import detect_cycle
from rulecat.detect_run_cycle import detect_run_cycle


run_max_depth = 64  # 깊이 백스톱


class Graph:
    def __init__(self, name: str) -> None:
        self.name = name
        self.rules: list[RuleMeta] = []
        self.roles: dict[str, str] = {}
        self.defeats: list[DefeatEdge] = []

    def rule(self, fn: Any) -> Rule:
        return self._register(fn, Strength.DEFEASIBLE, "rule")

    def counter(self, fn: Any) -> Rule:
        return self._register(fn, Strength.DEFEASIBLE, "counter")

    def except_(self, fn: Any) -> Rule:
        return self._register(fn, Strength.DEFEATER, "except")

    def _register(self, fn: Any, strength: Strength, role: str) -> Rule:
        rid = rule_id(fn, [])
        if rid in self.roles:
            raise ValueError(f"duplicate rule registration: {rid}")
        idx = len(self.rules)
        self.rules.append(RuleMeta(
            name=rid, qualifier=1.0, strength=int(strength),
            defeats=[], specs=[], fn=fn,
        ))
        self.roles[rid] = role
        return Rule(rid, self, idx, fn)

    def evaluate(self, ctx: Context, option: EvalOption | None = None) -> list[EvalResult]:
        if ctx is None:
            raise ValueError("ctx must not be None")
        results, *_ = self._evaluate(ctx, _resolve_option(option), full=False)
        return results

    def run(self, ctx: Context, option: EvalOption | None = None) -> RunResult:
        if ctx is None:
            raise ValueError("ctx must not be None")
        err = detect_run_cycle(self)  # 최상위 1회
        if err is not None:
            raise RuntimeError(f"toulmin: {err}")
        opt = _resolve_option(option)
        opt = EvalOption(method=opt.method, trace=False, duration=False)  # 강제 off
        return self._run_depth(ctx, opt, 0)

    def _run_depth(self, ctx: Context, opt: EvalOption, depth: int) -> RunResult:
        if depth > run_max_depth:
            raise RuntimeError(
                f"toulmin: run depth exceeded {run_max_depth} "
                "(possible runaway composition)"
            )
        results, active, verdict_cache, evidence = self._evaluate(ctx, opt, full=True)

        # 디스패치 전 1회 — 전 노드 TraceEntry를 등록 순서로 조립
        # (인덱스 i ↔ self.rules[i] 구조적 보장)
        entries: list[TraceEntry] = []
        for r in self.rules:
            entries.append(TraceEntry(
                name=short_name(r.name),
                role=self.roles.get(r.name, "rule"),
                activated=active.get(r.name, False),
                qualifier=r.qualifier,
                verdict=verdict_cache.get(r.name, 0.0),
                evidence=evidence.get(r.name),
                ground=ctx,
                specs=r.specs,
            ))

        tr = Trace(entries, ctx)

        for i, e in enumerate(entries):
            meta = self.rules[i]            # 인덱스 대응
            active_flag = e.activated and e.verdict > 0  # Active 하나만 발화
            if not active_flag:
                continue
            # (a) run_on 핸들러 — 먼저
            if meta.run_on is not None:
                try:
                    meta.run_on(e, tr)
                except Exception as exc:
                    raise RuntimeError(
                        f'run_on "{e.name}": {exc}'
                    ) from exc
            # (b) 실행 간선 — Active면 하위 그래프 Run (ctx 아래로, depth+1)
            if meta.run_graph is not None:
                try:
                    meta.run_graph._run_depth(ctx, opt, depth + 1)
                except Exception as exc:
                    raise RuntimeError(
                        f'run "{e.name}" → "{meta.run_graph.name}": {exc}'
                    ) from exc
        return RunResult(results=results, trace=tr)

    def _evaluate(
        self, ctx: Context, opt: EvalOption, full: bool
    ) -> tuple[list[EvalResult], dict[str, bool], dict[str, float], dict[str, Any]]:
        # Build eval maps
        fn_map: dict[str, Any] = {}
        qual_map: dict[str, float] = {}
        str_map: dict[str, int] = {}
        specs_map: dict[str, list[Any]] = {}
        edges: dict[str, list[str]] = {}

        for r in self.rules:
            fn_map[r.name] = r.fn
            qual_map[r.name] = r.qualifier
            str_map[r.name] = r.strength
            specs_map[r.name] = r.specs

        for d in self.defeats:
            if d.to not in edges:
                edges[d.to] = []
            edges[d.to].append(d.from_)

        # Cycle detection
        cycle_err = detect_cycle(edges)
        if cycle_err:
            raise RuntimeError(cycle_err)

        # Attacker set
        attacker_set: set[str] = set()
        for attackers in edges.values():
            for a in attackers:
                attacker_set.add(a)

        # Eval state
        ran: set[str] = set()
        active: dict[str, bool] = {}
        evidence: dict[str, Any] = {}
        verdict_cache: dict[str, float] = {}
        trace_entries: list[TraceEntry] = []
        err: list[Exception | None] = [None]  # mutable container for closure

        def is_warrant(name: str) -> bool:
            s = str_map.get(name, Strength.DEFEASIBLE)
            if s == Strength.DEFEATER:
                return False
            return name not in attacker_set

        def infer_role(id: str) -> str:
            if str_map.get(id) == Strength.DEFEATER:
                return "except"
            if id in attacker_set:
                return "counter"
            return "rule"

        def calc(id: str) -> float:
            if err[0]:
                return -1.0
            cached = verdict_cache.get(id)
            if cached is not None:
                return cached
            fn = fn_map.get(id)
            if fn is None:
                return -1.0
            if id not in ran:
                ran.add(id)
                try:
                    act, ev = fn(ctx, specs_map.get(id, []))
                    active[id] = act
                    evidence[id] = ev
                except Exception as e:
                    err[0] = RuntimeError(f'rule "{id}": {e}')
                    return -1.0
            if not active.get(id):
                return -1.0
            total = 0.0
            if str_map.get(id) != Strength.STRICT:
                for aid in edges.get(id, []):
                    total += (calc(aid) + 1.0) / 2.0
            q = qual_map.get(id, 1.0)
            v = 2.0 * (q / (1.0 + total)) - 1.0
            verdict_cache[id] = v
            return v

        def calc_trace(id: str, with_duration: bool) -> float:
            if err[0]:
                return -1.0
            cached = verdict_cache.get(id)
            if cached is not None:
                return cached
            fn = fn_map.get(id)
            if fn is None:
                return -1.0
            if id not in ran:
                ran.add(id)
                start = time.perf_counter() if with_duration else 0.0
                try:
                    act, ev = fn(ctx, specs_map.get(id, []))
                    active[id] = act
                    evidence[id] = ev
                except Exception as e:
                    err[0] = RuntimeError(f'rule "{id}": {e}')
                    return -1.0
                elapsed = time.perf_counter() - start if with_duration else None
                trace_entries.append(TraceEntry(
                    name=id,
                    role=infer_role(id),
                    activated=active.get(id, False),
                    qualifier=qual_map.get(id, 1.0),
                    evidence=evidence.get(id),
                    specs=specs_map.get(id, []),
                    duration=elapsed,
                ))
            if not active.get(id):
                return -1.0
            total = 0.0
            if str_map.get(id) != Strength.STRICT:
                for aid in edges.get(id, []):
                    total += (calc_trace(aid, with_duration) + 1.0) / 2.0
            q = qual_map.get(id, 1.0)
            v = 2.0 * (q / (1.0 + total)) - 1.0
            verdict_cache[id] = v
            return v

        def reset() -> None:
            ran.clear()
            active.clear()
            evidence.clear()
            verdict_cache.clear()
            trace_entries.clear()
            err[0] = None

        # Main evaluation loop
        results: list[EvalResult] = []
        for r in self.rules:
            if not is_warrant(r.name):
                continue
            if opt.trace:
                reset()
            verdict = (
                calc_trace(r.name, opt.duration) if opt.trace
                else calc(r.name)
            )
            if err[0]:
                raise err[0]
            if not active.get(r.name):
                continue
            result = EvalResult(
                name=short_name(r.name),
                verdict=verdict,
                evidence=evidence.get(r.name),
            )
            if opt.trace:
                result.trace = [
                    TraceEntry(
                        name=short_name(t.name),
                        role=t.role,
                        activated=t.activated,
                        qualifier=t.qualifier,
                        evidence=t.evidence,
                        specs=t.specs,
                        duration=t.duration,
                    )
                    for t in trace_entries
                ]
            results.append(result)

        if full:  # run이 trace를 강제 off하므로 calc(non-trace)만 탐
            for r in self.rules:
                if r.name not in ran:
                    calc(r.name)  # 미도달 노드까지 평가
            if err[0]:
                raise err[0]

        return results, active, verdict_cache, evidence


def _resolve_option(opt: EvalOption | None) -> EvalOption:
    if opt is None:
        return EvalOption()
    if opt.duration:
        opt = EvalOption(method=opt.method, trace=True, duration=True)
    return opt
